### ISSUE TITLE
Add Bootstrap 4.0.0-alpha.6

### ISSUE THREAD

**Issue by digibum** [MarkPieszak/aspnetcore-angular2-universal#288]
This might be simple for you guys, but i can't figure it out.
How to add Bootstrap v.4?

My package.json file contains:
`"bootstrap": "4.0.0-alpha.6",
    "bootstrap-sass": "4.0.0-alpha.6",`

Error:
`npm ERR! notarget No compatible version found: bootstrap-sass@4.0.0-alpha.6
npm ERR! notarget Valid install targets:
npm ERR! notarget 3.3.7, 3.3.6, 3.3.5, 3.3.4, 3.3.3, 3.3.2, 3.3.1, 3.3.0, 3.2.0, 3.1.1, 2.3.2`

Thank you all :)

**Comment by kyse** [comment]
I just use the sass that comes in the bootstrap 4 package.  Didn't realize they had a bootstrap-sass package for 4.0.  Either way I just grab the sass from their files from the main npm package.

You'll notice bootstrap 4 doesn't include icons anymore, so you'll probably want  font awesome as well.

in: Client/app/app.component.scss
```sass
/* Import Bootstrap & Fonts */
$fa-font-path: '~font-awesome/fonts/';
@import "~font-awesome/scss/font-awesome.scss";
@import "~bootstrap/scss/bootstrap.scss";
```

I don't remember exactly why I needed to do this: I think it was related to images in scss background, and something else.. but in case you run into another issue, this might help you.

in the webpack folder's webpack.common.js file:
Update your rules:
```typescript
  module: {
    rules: [
      { test: /\.ts$/, loader: '@ngtools/webpack' },
      { test: /\.css$/, loader: ['to-string-loader', 'css-loader', 'resolve-url-loader'] },
      { test: /\.html$/, loader: 'html-loader' },
      { test: /\.scss$/, loaders: ['to-string-loader', 'css-loader', 'resolve-url-loader', 'sass-loader?sourceMap'] },
      { test: /\.(woff2?|ttf|eot|svg)$/, loader: 'url-loader?limit=10000' },
      { test: /\.png$/, loader: 'url-loader?limit=25000&emitFile=false' }
    ]
  },
```
Might need to add a npm loader package.  I think resolve-url-loader is the new one, and it requires sass-loader to provide sourceMap.

You'll probably want to include the bootstrap js cdn in your Views/Shared/_Layout.cshtml so the bootstrap 4 navbar collapse functionality works properly.

Here's my updated CDN imports:
```html
        <script src="https://code.jquery.com/jquery-3.1.1.slim.min.js" integrity="sha384-A7FZj7v+d/sdmMqp/nOQwliLvUsJfDHW+k9Omg/a/EheAdgtzNs3hpfag6Ed950n" crossorigin="anonymous"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/tether/1.4.0/js/tether.min.js" integrity="sha384-DztdAPBWPRXSA/3eYEEUWrWCy7G5KFbe8fFjk5JAIxUYHKkDx6Qin1DkWx51bBrb" crossorigin="anonymous"></script>
        <script src="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0-alpha.6/js/bootstrap.min.js" integrity="sha384-vBWWzlZJ8ea9aCX4pEW3rVHjgjt7zpkNpZk+02D9phzyeVkE+jo0ieGizqPLForn" crossorigin="anonymous"></script>
        <script src="http://ajax.aspnetcdn.com/ajax/signalr/jquery.signalr-2.2.2.min.js"></script>
```

**Comment by digibum** [comment]
Thank you.
It's working :)

**Comment by donstarkey** [comment]
I hope you can pardon my ignorance on this matter.  I want to make use of the font-awesome library.

/* Import Bootstrap & Fonts */
$icon-font-path: '~bootstrap-sass/assets/fonts/bootstrap/';
@import "~bootstrap-sass/assets/stylesheets/bootstrap";
@import"~font-awesome/scss/font-awesome.scss";

then I modified webpack.common.js like you listed above
  module: {
    rules: [
      { test: /\.ts$/, loader: '@ngtools/webpack' },
      { test: /\.css$/, loader: ['to-string-loader', 'css-loader', 'resolve-url-loader'] },
      { test: /\.html$/, loader: 'html-loader' },
      { test: /\.scss$/, loaders: ['to-string-loader', 'css-loader', 'resolve-url-loader', 'sass-loader?sourceMap'] },
      { test: /\.(woff2?|ttf|eot|svg)$/, loader: 'url-loader?limit=10000' },
      { test: /\.png$/, loader: 'url-loader?limit=25000&emitFile=false' }
    ]
  },

However receiving the following error on npm run build:dev

![image](https://user-images.githubusercontent.com/16799754/26893798-04bc2964-4b82-11e7-9a08-089b4cced741.png)

I am not very familiar with the client side, so really at a loss right now.  By any chance is there any detailed documentation regarding the use of the files in the solution?  It would really help me to understand how everything fits together.  What can I do to simply have font-awesome and the alertify js library included in this solution.  At a loss.

**Comment by donstarkey** [comment]
I just added font-awesome directly to the head of the _Layout.cshtml.  Now images are showing up, but I would like to do this in an optimized fashion.  Any help is greatly appreciated!!

**Comment by isaac2004** [comment]
@kyse are you using ngx or ng2 bootstrap? My reason for asking is that I do not think animations are supported yet in ngx. So if you have a version of the nav that is responsive using ngx and bs4, I would love to see it!

**Comment by kyse** [comment]
I've never seen an implementation of the navbar worth a damn from any angularized bootstrap.  I always roll my own nav since they always tend to be slightly different in one way or another.

I'm using bootstrap 4 sass, imported the bs 4 js cdn for collapse functionality.  Using ng-bootstrap 1.0.0-alpha26 for angularized bootstrap in various places which seems to work with bs4 css, but like I said, not for navbar.

Per [Navbar - Bootstrap](https://v4-alpha.getbootstrap.com/components/navbar/), "Navbars are responsive by default, but you can easily modify them to change that. Responsive behavior depends on our Collapse JavaScript plugin."  I take that to mean I need their js for collapse to work.  Perhaps not but it didn't seem to work till I imported it.

For clarity, here's my implementation of a navbar I'm currently using.  It achieves specific goals for a specific site though, so wont look like any stock navbar.

### Navbar Component
The navbar parses the routes for entries where data.navbar = true, and uses those to build out the top navbar.  I haven't added detailed auth restrictions on it yet since I haven't found a way to handle executing an authguard to determine if specific requirements are met from outside the router actions itself.  Provides 2 different sizes for the navbar, then collapses to a centered list after a breakpoint.

#### navbar.component.html
```html
<div class="navbar navbar-toggleable-sm navbar-inverse bg-inverse p-0">
    <button *ngIf="menuItems.length" class="navbar-toggler navbar-toggler-right" type="button" data-toggle="collapse" data-target="#sitename-navbar-main" aria-controls="sitename-navbar-main" aria-expanded="false" aria-label="Toggle navigation">
        <span class="navbar-toggler-icon"></span>
    </button>
    <a class="navbar-brand navbar-toggler-left" [routerLink]="['/']">
        <img src="/images/logos/128.png" class="hidden-md-down" alt="SiteNameLogo">
        <img src="/images/logos/64.png" class="hidden-sm-down hidden-lg-up" alt="SiteName Logo">
        <span class="hidden-md-up">
            <img src="/images/logos/32.png" class="d-inline-block align-top" alt="SiteName Logo">
            <span class="logo-text">SiteName</span>
        </span>
    </a>
    <div class="container-fluid">
        <div class="row justify-content-center collapse navbar-collapse" id="sitename-navbar-main">
            <ul class="navbar-nav mt-2 mt-md-0">
                <li *ngFor="let menuItem of menuItems" class="navbar-item" [class.hidden-md-up]="menuItem.collapsedOnly" [routerLinkActive]="['active']">
                    <a class="nav-link" [routerLink]="[menuItem.path]">
                        <span class="fa fa-{{menuItem.icon}}" [translate]="menuItem.i18n">{{menuItem.title}}</span>
                    </a>
                </li>
            </ul>
            <div *ngIf="isAuthenticated" class="justify-content-center search input-group">
                <span class="input-group-addon">
                    <i class="fa fa-search"></i>
                </span>
                <input type="text" class="form-control" placeholder="{{'searchplaceholder' | translate}}" autocomplete="off" spellcheck="false" dir="auto"/>
            </div>
        </div>
    </div>
    <navAccountWidget></navAccountWidget>
</div>
```

#### navbar.component.css
```css
.logo-text {
    color: #fff;
    font-weight: 300;
}

.navbar .container-fluid .row {
    margin-left: 0 !important;
    margin-right: 0 !important;
}

.navbar ul li a,
.navbar ul li.link-active a:hover,
.navbar ul li.link-active a:focus {
    color: #b9b9b9;
    font-weight: 100;
    text-align: center;
}

.navbar ul li a:focus {
    background-color: transparent;
    text-decoration: none;
}

.navbar ul li a:hover {
    background-color: #555555;
    text-decoration: none;
}

@media (min-width: 768px) {
    .navbar {
        min-height: 151px;
    }

    .navbar .navbar-brand {
        top: 0;
        padding: 22px 15px;
        margin-left: -15px;
    }

    .navbar ul li {
        margin: 6px;
    }

    .navbar ul li a .fa {
        display: inline;
    }

    .navbar ul li a .fa:before {
        font-size: 35px;
        display: block;
        margin-bottom: 1px;
        box-sizing: border-box;
    }
}

@media(min-width: 992px) {
    .navbar ul li a,
    .navbar ul li.link-active a:hover,
    .navbar ul li.link-active a:focus {
        border-radius: 6px;
        padding: 15px 10px 5px;
        min-height: 76px;
        min-width: 84px;
        margin: 20px 10px 5px;
    }
}

@media (min-width: 768px) and (max-width: 991px) {
    .navbar ul li a,
    .navbar ul li.link-active a:hover,
    .navbar ul li.link-active a:focus {
        border-radius: 6px;
        padding: 5px 0 5px;
        min-height: 76px;
        min-width: 64px;
        margin: 20px 5px 5px;
    }

}

@media (max-width: 767px) {
    .navbar { min-height: 58px; }

    .navbar .container-fluid { margin-top: 40px; }

    .navbar .navbar-brand,
    .navbar .navbar-toggler { top: 1rem; }

    .navbar ul li a .fa:before { margin-right: 1rem; }
}

.search.input-group .input-group-addon,
.search.input-group input {
    background-color: #333;
}

.search.input-group .input-group-addon,
.search.input-group input {
    border-color: #333;
    color: #ccc;
}

.search.input-group input:focus {
    border-color: #f08080;
}

.navbar {
    background-color: transparent !important;
}
```

#### navbar.component.ts
```typescript
import { Component, OnInit, OnDestroy } from '@angular/core';
import { Router, Route } from '@angular/router';
import { Subscription } from 'rxjs/Subscription';

import { AuthService } from '../../shared/auth.service';

export class NavMenuItem {
    title: string;
    i18n: string;
    icon: string;
    path: string;
    order: number;
    navbar: boolean;
    collapsedOnly: boolean;

    constructor(route: Route) {
        this.path = route.path;
        this.icon = route.data['icon'];
        this.title = route.data['title'];
        this.i18n = route.data['i18n'];
        this.order = route.data['order'] || 0;
        this.navbar = route.data['navbar'];
        this.collapsedOnly = route.data['collapsedOnly'] || false;
    }
}

@Component({
    selector: '[navMenu]',
    templateUrl: './navmenu.component.html',
    styleUrls: ['./navmenu.component.css']
})
export class NavMenuComponent implements OnInit, OnDestroy {
    menuItems: NavMenuItem[] = [];
    isAuthenticated: boolean;
    private $isAuthenticated: Subscription;

    constructor(
        private router: Router,
        private authService: AuthService,
    ) {
        this.isAuthenticated = this.authService.isAutheneticated();
        this.$isAuthenticated = this.authService.$isAutheneticated().subscribe(authenticated => this.authenticationChangedHandler(authenticated));
        this.createMenuItems();
    }

    ngOnInit() {
    }

    ngOnDestroy() {
        this.$isAuthenticated.unsubscribe();
    }

    private createMenuItems() {
        this.menuItems = this.router.config
            .filter((config: Route) => config.data && config.data['navbar'] && (this.isAuthenticated || !config.canActivate || !config.canActivate.length))
            .map(config => new NavMenuItem(config))
            .sort((a, b) => a.order - b.order);
    }

    private authenticationChangedHandler(authenticated: boolean) {
        this.isAuthenticated = authenticated;
        this.createMenuItems();
    }
}
```

### Subnav Component
I also have a subnav component for displaying a responsive subnav tab/pills hybrid using bootstrap scss mixins to create my own merged class of the two for displaying child routes at the top of the main page content (inside the root router-outlet, there should be a sub router-outlet for these child navigation navs).  Could probably expand this to do some collapsing as well, but for my needs, well, no need.  Would be nice however to figure out how to tree shake scss to get rid of all the unused imported bootstrap scss.  I probably just need to do more selective importing somehow, not sure tree shaking scss is even a reality.

#### subnav.component.html
```html
<nav>
    <ul class="nav nav-tabs-to-pills-md nav-justified">
        <li *ngFor="let tab of content" class="nav-item">
            <a [id]="tab.id"
               class="nav-link"
               [class.active]="tab.id === activeId"
               [class.disabled]="tab.disabled"
               href="#"
               (click)="!!onTabSelected(tab.id)"
               role="tab"
               [attr.tabindex]="(tab.disabled ? '-1' : undefined)"
               [attr.aria-expanded]="tab.id === activeId"
               [attr.aria-disabled]="tab.disabled"
               [translate]="tab.i18n">{{tab.title}}</a>
        </li>
    </ul>    
</nav>
```

#### subnav.component.scss
```scss
@import "~bootstrap/scss/bootstrap.scss";

.nav-tabs-to-pills {
    @each $breakpoint in map-keys($grid-breakpoints) {
        $next: breakpoint-next($breakpoint, $grid-breakpoints);
        $infix: breakpoint-infix($next, $grid-breakpoints);
        &#{$infix} {
            @include media-breakpoint-down($breakpoint) {
                flex-direction: column !important;

                .nav-link {
                    @include border-radius($nav-pills-border-radius);
                }

                .nav-link.active,
                .nav-item.show .nav-link {
                    color: $nav-pills-active-link-color;
                    cursor: default;
                    background-color: $nav-pills-active-link-bg;
                }
            }

            @include media-breakpoint-up($next) {
                flex-direction: row !important;
                border-bottom: $nav-tabs-border-width solid $nav-tabs-border-color;

                .nav-item {
                    margin-bottom: -$nav-tabs-border-width;
                }

                .nav-link {
                    border: $nav-tabs-border-width solid transparent;
                    @include border-top-radius($nav-tabs-border-radius);

                    @include hover-focus {
                        border-color: $nav-tabs-link-hover-border-color $nav-tabs-link-hover-border-color $nav-tabs-border-color;
                    }

                    &.disabled {
                        color: $nav-disabled-link-color;
                        background-color: transparent;
                        border-color: transparent;
                    }
                }

                .nav-link.active,
                .nav-item.show .nav-link {
                    color: $nav-tabs-active-link-hover-color;
                    background-color: $nav-tabs-active-link-hover-bg;
                    border-color: $nav-tabs-active-link-hover-border-color $nav-tabs-active-link-hover-border-color $nav-tabs-active-link-hover-bg;
                }

                .dropdown-menu {
                    // Make dropdown border overlap tab border
                    margin-top: -$nav-tabs-border-width;
                    // Remove the top rounded corners here since there is a hard edge above the menu
                    @include border-top-radius(0);
                }
            }
        }
    }
}
```

#### subnav.component.ts
```typescript
import { Component, OnInit, OnDestroy, Input, Output, EventEmitter } from '@angular/core';
import { Router, Route, ActivatedRoute } from '@angular/router';
import { NgbTabChangeEvent } from '@ng-bootstrap/ng-bootstrap';

import { Subscription } from 'rxjs/Subscription';

export class SubNavContent {
    id: string;
    title: string;
    i18n: string;
    icon: string;
    path: string;
    disabled: boolean;

    constructor(route: Route) {
        this.id = route.path;
        this.title = route.data.title;
        this.i18n = route.data.i18n;
        this.icon = route.data.icon;
        this.path = route.path;
        this.disabled = route.data.disabled;
    }
}

export interface SubNavTabChangedEvent {
    activeId: string;
    nextId: string;
    preventDefault: () => void;
}

@Component({
    selector: 'subNav',
    templateUrl: './subnav.component.html',
    styleUrls: ['./subnav.component.scss']
})
export class SubNavComponent implements OnInit, OnDestroy {
    @Input() content: SubNavContent[];
    @Input() type: 'tabs' | 'pills' = 'tabs';
    @Input() justify: 'start' | 'center' | 'end' = 'center';
    @Output() tabChanged = new EventEmitter<SubNavTabChangedEvent>();
    activeId: string;
    
    constructor(
        private route: ActivatedRoute,
        private router: Router
    ) {
    }

    ngOnInit() {
        this.activeId = this.getInitialTab();
    }

    ngOnDestroy() {
    }

    getInitialTab(): string {
        if (!this.content || !this.content.length)
            return '';
        var needsNavigation = !this.route.children.length;
        if (!needsNavigation) {
            return this.createTabId(this.route.children[0].routeConfig.path);
        } else {
            if (!this.route.routeConfig.children.length) {
                console.log('No children paths available.');
                return '';
            } else {
                let activeId = this.createTabId(this.route.routeConfig.children[0].path);
                this.router.navigate([this.getRelativePath(activeId)], { relativeTo: this.route });
                return activeId;
            }
        }
    }

    private createTabId(targetPath: string): string {
        //return this.route.routeConfig.path + "." + targetPath;
        return targetPath;
    }

    private getRelativePath(tabId: string): string {
        return tabId.replace(this.route.routeConfig.path + '.', '');
    }

    public onTabSelected(tabId: string) {
        let selectedTab = this.getTabById(tabId);
        if (selectedTab && !selectedTab.disabled && this.activeId !== selectedTab.id) {
            let defaultPrevented = false;
            this.tabChanged.emit({
                activeId: this.activeId,
                nextId: selectedTab.id,
                preventDefault: () => {
                    defaultPrevented = true;
                }
            });
            if (!defaultPrevented) {
                this.activeId = selectedTab.id;
                this.router.navigate([selectedTab.path], { relativeTo: this.route });
            }
        }
    }

    private getTabById(id: string): SubNavContent {
        let tabs = this.content.filter(tab => tab.id === id);
        return tabs.length ? tabs[0] : null;
    }
}
```
